File: 2022/python/day_6.py
#!/usr/bin/env python3

import os
from collections import deque

INPUT_FILEPATH = os.path.join(os.path.dirname(__file__), "../inputs/day_6.txt")


def look_for_n_distinct_chars(n, message):
    window = deque(message[:n])
    idx = n-1
    while True:
        if len(set(window)) == n:
            break
        idx += 1
        window.popleft()
        window.append(message[idx])
    return idx+1


def main():

    with open(INPUT_FILEPATH) as input:
        message = input.read()
        
    print("Position for 4 chars is", look_for_n_distinct_chars(4, message))
    print("Position for 14 chars is", look_for_n_distinct_chars(14, message))


if __name__ == "__main__":
    main()

File: 2022/rust/src/bin/day_6.rs
use std::char;
use std::collections::HashSet;
use std::fs::read_to_string;
use std::path::Path;

const INPUT_PATH_REL: &str = "../../../inputs/day_6.txt";
const SRC_PATH: &str = concat!(env!("CARGO_MANIFEST_DIR"), "/", file!());

fn look_for_n_distinct_chars(n: usize, message: &[char]) -> Option<usize> {
    for (idx, window) in message.windows(n).enumerate() {
        if HashSet::<&char>::from_iter(window).len() == n {
            return Some(idx + n);
        }
    }
    None
}

fn main() {
    let input_path = Path::new(SRC_PATH).parent().unwrap().join(INPUT_PATH_REL);
    let input = read_to_string(input_path).unwrap();

    let chars = input.chars().collect::<Vec<_>>();
    println!(
        "Position for 4 chars is {}",
        look_for_n_distinct_chars(4, &chars).unwrap()
    );
    println!(
        "Position for 14 chars is {}",
        look_for_n_distinct_chars(14, &chars).unwrap()
    );
}
